nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java

```diff
@@ -181,133 +186,220 @@ public IcebergTable table(TableIdentifier tableIdentifier) {
               namespace, getRef().getName()),
           e);
     }
   }
 
   public boolean setProperties(Namespace namespace, Map<String, String> properties) {
-    try {
-      withReference(
-              getApi()
-                  .updateProperties()
-                  .namespace(org.projectnessie.model.Namespace.of(namespace.levels()))
-                  .updateProperties(properties))
-          .update();
-      refresh();
-      // always successful, otherwise an exception is thrown
-      return true;
-    } catch (NessieNamespaceNotFoundException e) {
-      throw new NoSuchNamespaceException(e, "Namespace does not exist: %s", namespace);
-    } catch (NessieNotFoundException e) {
-      throw new RuntimeException(
-          String.format(
-              "Cannot update properties on Namespace '%s': ref '%s' is no longer valid.",
-              namespace, getRef().getName()),
-          e);
-    }
+    return updateProperties(namespace, props -> props.putAll(properties));
   }
 
   public boolean removeProperties(Namespace namespace, Set<String> properties) {
+    return updateProperties(namespace, props -> props.keySet().removeAll(properties));
+  }
+
+  private boolean updateProperties(Namespace namespace, Consumer<Map<String, String>> action) {
+    getRef().checkMutable();
+    ContentKey key = ContentKey.of(namespace.levels());
     try {
-      withReference(
-              getApi()
-                  .updateProperties()
-                  .namespace(org.projectnessie.model.Namespace.of(namespace.levels()))
-                  .removeProperties(properties))
-          .update();
-      refresh();
+      commitRetry(
+          "update namespace " + key,
+          true,
+          commitBuilder -> {
+            Map<ContentKey, Content> contentMap =
+                api.getContent().reference(getReference()).key(key).get();
+            Content existing = contentMap.get(key);
+            if (!(existing instanceof org.projectnessie.model.Namespace)) {
+              throw new NessieContentNotFoundException(key, getReference().getName());
+            }
+            org.projectnessie.model.Namespace oldNamespace =
+                (org.projectnessie.model.Namespace) existing;
+            Map<String, String> newProperties = Maps.newHashMap(oldNamespace.getProperties());
+            action.accept(newProperties);
+            org.projectnessie.model.Namespace updatedNamespace =
+                org.projectnessie.model.Namespace.builder()
+                    .from(oldNamespace)
+                    .properties(newProperties)
+                    .build();
+            commitBuilder.operation(Operation.Put.of(key, updatedNamespace));
+            return commitBuilder;
+          });
       // always successful, otherwise an exception is thrown
       return true;
-    } catch (NessieNamespaceNotFoundException e) {
+    } catch (NessieReferenceConflictException e) {
+      NessieConflictHandler.handleSingle(
```

Comment:
this error handling pattern seems a bit weird. `handleSingle()` returns a boolean, which we ignore here and fall-through to then throw a `RuntimeException`. In another case we actually handle the boolean.

Why not have `handleSingle()` take an optional exception that will be thrown in case nothing matches?

Reply:
You are right, I will refactor a bit more. I think we just need an utility method that simply returns an `Optional<Conflict>`.